text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+
+### Restricting access to an S3 Origin
+
+CloudFront provides two ways to send authenticated requests to an Amazon S3 origin: origin access control (OAC) and origin access identity (OAI).
+OAC is the recommended method and OAI is considered legacy (see [Restricting access to an Amazon Simple Storage Service origin](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html)).
+Following AWS best practices, it is recommended you set the feature flag `@aws-cdk/aws-cloudfront:useOriginAccessControlByDefault` to `true` to use OAC by
+default when creating new origins.
+
+For an S3 bucket that is configured as a standard S3 bucket origin (not as a website endpoint), when the above feature flag is enabled the `S3Origin`
+construct will automatically create an OAC and grant it access to the underlying bucket.
+
+> [Note](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html): When you use OAC with S3
+bucket origins you must set the bucket's object ownership to Bucket owner enforced, or Bucket owner preferred (only if you require ACLs).
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket', {
+  objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
+});
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: { 
+    origin: new origins.S3Origin(myBucket) // Automatically creates an OAC
+  },
+});
+```
+
+Alternatively, a custom origin access control can be passed to the S3 origin:
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket');
+const myOAC = new cloudfront.OriginAccessControl(this, 'myOAC', {
+  description: 'Origin access control for S3 origin',
+  originAccessControlOriginType: cloudfront.OriginAccessControlOriginType.S3,
+});
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: { 
+    origin: new origins.S3Origin(myBucket, {
+      originAccessControl: myOAC
+    }),
+  },
+});
+```
+
+Alternatively, an existing origin access control can be imported:
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket');
+const importedOAC = cloudfront.OriginAccessControl.fromOriginAccessControlAttributes(this, 'myImportedOAC', {
+  originAccessControlId: 'ABC123ABC123AB',
+  originAccessControlOriginType: cloudfront.OriginAccessControlOriginType.S3,
+});
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: { 
+    origin: new origins.S3Origin(myBucket, {
+      originAccessControl: importedOAC
+    }),
+  },
+});
+```
+
+If the feature flag is not enabled (i.e. set to `false`), an origin access identity will be created by default.
+
+#### Using OAC for a SSE-KMS encrypted S3 origin
+
+If the objects in the S3 bucket origin are encrypted using server-side encryption with
+AWS Key Management Service (SSE-KMS), the OAC must have permission to use the AWS KMS key.
+A statement needs to be added to the KMS key policy to give the OAC permission to use the KMS key.
```

Comment:
But, we should only do this on non-imported keys for now.

Reply:
I agree that we should do this for them. @TheRealAmazonKendra can you explain why we shouldn't do this for imported keys? I agree with you that we should not, and I think that's because an imported key likely already has an existing policy associated with it (if it didn't, then it wouldn't be in use, and why bother importing it at all?) so it's best to not mess with it.